solidity/src/examples/PayMe.sol

```diff
@@ -0,0 +1,46 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.13;
+
+import {RollupOrders} from "zenith/src/orders/RollupOrders.sol";
+import {SignetStd} from "../SignetStd.sol";
+
+/// @notice This contract provides a modifier that allows functions to be gated
+///         by requiring a payment of a specified amount of native asset.
+abstract contract PayMe is SignetStd {
+    /// @notice This modifier crates an order with no input, that pays the
+    ///         specified amount of native asset to the contract. It can be used
+    ///         to gate access to payment-gate functions.
+    modifier payMe(uint256 amount) {
+        demandPayment(NATIVE_ASSET, amount);
+
+        _;
+    }
+
+    /// @notice Creates an order that demands payment of the specified amount
+    ///         of the specified asset to the contract.
+    /// @dev This is useful for cases where the payment should go to the
+    ///      contract itself, such as for fees or service charges.
+    /// @param asset The address of the asset to be paid.
+    /// @param amount The amount of the asset to be paid.
+    function demandPayment(address asset, uint256 amount) internal {
+        demandPaymentTo(asset, amount, address(this));
+    }
+
+    /// @notice Creates an order that demands payment of the specified amount
+    ///         of the specified asset to a specific recipient.
+    /// @dev This is useful for cases where the payment should go to a different
+    ///      address, such as a treasury or a specific user.
+    /// @param asset The address of the asset to be paid.
+    /// @param amount The amount of the asset to be paid.
+    /// @param recipient The address that will receive the payment.
+    function demandPaymentTo(address asset, uint256 amount, address recipient) internal {
+        RollupOrders.Output[] memory outputs = new RollupOrders.Output[](1);
+        outputs[0] = makeRollupOutput(asset, amount, recipient);
+
+        ORDERS.initiate(
+            type(uint256).max, // this is equivalent to no deadline
```

Comment:
x-post, but - in the case of a contract-initiated Order, can we always put `deadline = block.timestamp`? the function can’t be called, the transaction cannot run, unless the Order has _already been_ fulfilled.